File: c#/src/Sodium/Sodium/Cleanup.cs
using System;

namespace Sodium
{
    /// <summary>
    ///     An object which allows for arbitrary cleanup code to safely run when this object is garbage collected.
    /// </summary>
    public class Cleanup
    {
        // ReSharper disable once PrivateFieldCanBeConvertedToLocalVariable
        // ReSharper disable once NotAccessedField.Local
        private Stream<Unit> stream;

        public Cleanup(Action cleanup)
        {
            Stream<Unit> stream = Stream.Never<Unit>();
            stream.AttachListener(Listener.Create(cleanup));

            this.stream = stream;
        }

        /// <summary>
        ///     Force the cleanup to happen now rather than waiting for this object to be garbage collected.
        /// </summary>
        public void CleanupNow()
        {
            this.stream = null;
        }
    }
}

File: f#/src/SodiumFRP.FSharp/_Root/Cleanup.fs
namespace SodiumFRP.FSharp

type Cleanup (cleanup) =
    let mutable stream = Some (Stream<unit>().AttachListener <| Listener.fromAction cleanup)
    member internal __.CleanupNow () = stream <- None
    member internal __.Stream = stream // prevent GC

module Cleanup =
    let cleanupNow (cleanup : Cleanup) = cleanup.CleanupNow ()